### PR TITLE
Return statistics with runQuery; add statistics to BigQuery

### PR DESCRIPTION
**Pull request by lukesonnet**
### Features and Changes

* Adds statistics to query responses
* Adds statistics from BQ client

### Testing

* Metric value query with Postgres (works, no stats)
* Metric value query with BQ (works, stats in mongo, properly shows hitting partitions)
Screenshot 2023-08-25 at 1 32 00 PM

* past experiment query with Postgres (works, no stats)
* past experiment query with BQ (works, stats in mongo)
Screenshot 2023-08-25 at 1 32 25 PM

* experiment metric value queries with Postgres (works, no stats)
* experiment metric value queries with BQ (works, stats in mongo)
Screenshot 2023-08-25 at 2 08 14 PM

* tested some other random queries as well (e.g. metric modal test queries) and no bugs

Possible fast-follows:
* expose data to users in query modal
* aggregate data for overall runtime stats etc.

### CODE REVIEW

File: packages/back-end/src/integrations/BigQuery.ts [hunk 1]
```diff
@@ -47,7 +48,16 @@ export default class BigQuery extends SqlIntegration {
...
     const [rows] = await job.getQueryResults();
-    return rows;
+    const statistics = {
+      executionDurationMs: job.metadata.statistics.finalExecutionDurationMs,
+      totalSlotMs: job.metadata.statistics.totalSlotMs,
+      bytesProcessed: job.metadata.statistics.totalBytesProcessed,
+      bytesBilled: job.metadata.statistics.query.totalBytesBilled,
+      warehouseCachedResult: job.metadata.statistics.query.cacheHit,
+      partitionsUsed:
+        job.metadata.statistics.query.totalPartitionsProcessed > 0 ?? undefined,
```
Comment (jdorn) [inline packages/back-end/src/integrations/BigQuery.ts:58]: You have the type of this specified as a number, but are setting it to a boolean here

File: packages/back-end/src/types/Integration.ts [hunk 2]
```diff
@@ -188,6 +191,25 @@ export type ExperimentMetricQueryResponse = {
   covariate_sum_squares?: number;
   main_covariate_sum_product?: number;
 }[];
+export type QueryResponse = {
+  // eslint-disable-next-line
+  rows: any[];
+  statistics?: QueryStatistics;
+};
```
Comment (jdorn) [inline packages/back-end/src/types/Integration.ts:198]: Better to use generics here.

```ts
// Add a named generic 'Row' here
export type QueryResponse<Row> = {
  rows: Row[];
  statistics?: QueryStatistics;
}

// Now, the other definitions become much simpler
export type MetricValueQueryResponse = QueryResponse<MetricValueQueryResponseRow>;
```

File: packages/back-end/src/queryRunners/QueryRunner.ts [hunk 3]
```diff
@@ -264,7 +265,9 @@ export abstract class QueryRunner<
   >(
     name: string,
     query: string,
-    run: (query: string) => Promise<Rows>,
+    run: (
+      query: string
+    ) => Promise<{ statistics: QueryStatistics; rows: Rows }>,
```
Comment (jdorn) [inline packages/back-end/src/queryRunners/QueryRunner.ts:270]: `statistics` here should be optional

File: packages/back-end/src/services/experiments.ts [hunk 4]
```diff
@@ -477,6 +477,7 @@ export async function createSnapshot({
     snapshotSettings: data.settings,
     variationNames: experiment.variations.map((v) => v.name),
     metricMap,
+    queryParentId: snapshot.id,
```
Comment (jdorn) [inline packages/back-end/src/services/experiments.ts:480]: I don't think this is needed

File: packages/back-end/src/types/Integration.ts (hunk 2 above)
Comment (lukesonnet) [reply packages/back-end/src/types/Integration.ts:199]: I knew there had to be a way to do this, thank you so much!

File: packages/back-end/src/integrations/BigQuery.ts [hunk 5]
```diff
@@ -47,7 +48,21 @@ export default class BigQuery extends SqlIntegration {
       useLegacySql: false,
     });
     const [rows] = await job.getQueryResults();
-    return rows;
+    const [metadata] = await job.getMetadata();
```
Comment (lukesonnet) [inline packages/back-end/src/integrations/BigQuery.ts:51]: @jdorn , as mentioned on slack, I'm uncertain about this. Before, getting the metadata sometimes would return the "RUNNING" state and no statistics, even though it was below `await job.getQueryResults()`. At least in my test case where I could reliably get the "RUNNING" state, using `await job.getMetadata()` seems to have resolved the issue and it returns the "DONE" metadata instead.

Would like your advice here!

Comment (jdorn) [reply packages/back-end/src/integrations/BigQuery.ts:51]: If I were to guess, the `job.metadata` property is not meant to be accessed directly.  So using the `getMetadata()` function is likely the proper way to do it.

**Review by jdorn — APPROVED** [review]

**Pull request merged**